implement tlsRigour for AMQP broker connections

petersilva opened the pull request:
close #1525 

To honour tlsRigour,  need to construct an ssl argument for amqp Connection routine.

Second try at this problem.  Worth testing.  I tested a bit, and it worked when I faked hostnames, but now sure if it is enough to solve the entire problem.

petersilva committed:
implement tlsRigour for py-amqp

need to construct an ssl argument for amqp Connection routine.

reidsunderland commented:
I tested with a self-signed cert. It works perfectly on Ubuntu 22 (Python 3.10, amqp 5.0.9):

- strict and normal (default) complain about the self-signed cert and refuse to connect to the broker
- lax connects successfully without any complaint

 On Ubuntu 18 (Python 3.6, amqp 2.2.2):

- ✅ `strict` refuses to connect: `sarracenia.moth.amqp getSetup failed connection to amqps://bunnymaster@edcm-dirt22-4.edc-mtl.ec.gc.ca/: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:852)`
- ❓`normal` connects without any complaint
- _`lax` refuses to connect_: `sarracenia.moth.amqp getSetup failed connection to amqps://bunnymaster@edcm-dirt......ca/: Cannot set verify_mode to CERT_NONE when check_hostname is enabled.`

I tried adding `'check_hostname': False` to the args but that caused a different error, `_wrap_socket_sni() got an unexpected keyword argument 'check_hostname'`.

I don't really think it's a big deal, Python 3.6 is obsolete and it seems like the `normal` mode works fine even with certificates that would be rejected by `strict` mode. So I'm okay with approving and merging this if nobody else has concerns.

reidsunderland commented:
I also tried commenting out the `'server_hostname': None` line but it didn't make any difference either

petersilva commented:
I don't have access to 18.04... if someone has some time here are some things somebody can check:

in the lax scheme:

```
+            if self.o['tlsRigour'] == 'lax':
+                sslarg = {
+                    'cert_reqs' : ssl.CERT_NONE,
+                    'ssl_version':  ssl.PROTOCOL_TLS_CLIENT,
+                    'server_hostname': None
+               }

```
try adding:
```
                         'check_hostname': False,
```

It might work... If not then we need some idea of what they were expecting back then...
to get some documentation, maybe go on an 18.04 machine, and do:
```

python
import amqp
help(amqp.Connection)   # in the long text it should say look for help via the following:
help(amqp.transport.SSLTransport)

```

Maybe it will say what the fields are in the old version of py-amqp?

reidsunderland commented:
I tried adding `'check_hostname': False` but it gave a different error: 
`_wrap_socket_sni() got an unexpected keyword argument 'check_hostname'`

The only relevant info I see in `help(amqp.Connection)` is:

```
 |  The 'ssl' parameter may be simply True/False, or for Python >= 2.6
 |  a dictionary of options to pass to ssl.wrap_socket() such as
 |  requiring certain certificates.
```

`help(amqp.transport.SSLTransport)` doesn't address the ssl parameter, it just show that it exists in the signature and is None by default.

```
Help on function wrap_socket in module ssl:

wrap_socket(sock, keyfile=None, certfile=None, server_side=False, cert_reqs=<VerifyMode.CERT_NONE: 0>, ssl_version=<_SSLMethod.PROTOCOL_TLS: 2>, ca_certs=None, do_handshake_on_connect=True, suppress_ragged_eofs=True, ciphers=None)
```

So `cert_reqs`, `ssl_version` are both supported, but `server_hostname` and `check_hostname` aren't present.

petersilva commented:
I read a bit... can you replace PROTOCOL_TLS_CLIENT, with just PROTOCOL_TLS   on ubuntu 18.04? 
```
+                sslarg = {
+                    'cert_reqs' : ssl.CERT_NONE,
+                    'ssl_version':  ssl.PROTOCOL_TLS_CLIENT,
+                    'server_hostname': None
+                }
```
I was reading that PROTOCOL_TLS_CLIENT *enables CERT_REQUIRED and check_hostname by default.*  ... the older PROTOCOL_TLS doesn't (but is deprecated in newer OS.)

reidsunderland commented:
PROTOCOL_TLS works!

petersilva commented:
so if you have PROTOCOL_TLS on modern OS, does it still work? do we just change that for everyone?
Or do we need that for older versions and _CLIENT for the newer ones?

reidsunderland commented:
I tried PROTOCOL_TLS on Ubuntu 22.04/Python 3.10 and it worked fine there as well. I guess it would be okay to use it everywhere, and hopefully by the time it gets removed from Python we won't be using/supporting Python 3.6 anymore

petersilva committed:
tlsRigour lax needs PROTOCOL_TLS #1533

@reidsunderland tested it, and on ubuntu 18.04 need to use PROTOCOL_TLS to turn check_hostname off.
See discussion in PR #1533 for issue #1525

petersilva commented:
okey dokey...  I added the patch, please bless if happy.

reidsunderland reviewed (approved):
The latest changes are confirmed to work on old Python (3.6) and new.
